### PR TITLE
add node group roles in controllers

### PR DESCRIPTION
**Pull request by Thiblef**
[OO-FEATURE] Add node specifics roles

https://github.com/open-orchestra/open-orchestra-model-bundle/pull/514

### CODE REVIEW

File: ApiBundle/Controller/NodeController.php [hunk 1]
```diff
@@ -90,18 +99,19 @@ public function showOrCreateAction(Request $request, $nodeId, $errorNode)
...
      *
-     * @Config\Security("is_granted('ROLE_ACCESS_DELETE_NODE')")
-     *
      * @return Response
      */
-    public function deleteAction($nodeId)
+    public function deleteAction(Request $request, $nodeId)
     {
         $siteId = $this->get('open_orchestra_backoffice.context_manager')->getCurrentSiteId();
+        $node = $this->findOneNode($nodeId, $request->get('language'), $siteId);
```
Comment (itkg-alavieille) [inline ApiBundle/Controller/NodeController.php]: why not uses parameter $nodes to get one nodes ?

File: ApiBundle/Controller/NodeController.php [hunk 2]
```diff
@@ -228,7 +237,11 @@ public function changeStatusAction(Request $request, $nodeMongoId)
      */
     public function listStatusesForNodeAction($nodeMongoId)
     {
+        /** @var NodeInterface $node */
         $node = $this->get('open_orchestra_model.repository.node')->find($nodeMongoId);
+        if (!$node->getNodeType() === NodeInterface::TYPE_ERROR) {
+            $this->denyAccessUnlessGranted(TreeNodesPanelStrategy::ROLE_ACCESS_TREE_NODE, $node);
```
Comment (itkg-alavieille) [inline ApiBundle/Controller/NodeController.php]: why no check on the nodes of type error ?

File: BackofficeBundle/Controller/NodeController.php [hunk 3]
```diff
@@ -54,19 +54,21 @@ public function formAction(Request $request, $id)
...
-     * @Config\Security("is_granted('ROLE_ACCESS_CREATE_NODE')")
-     *
      * @return Response
      */
     public function newAction(Request $request, $parentId)
     {
+        $parentNode = $this->get('open_orchestra_model.repository.node')->findOneByNodeId($parentId);
+
+        $creationRole = $parentNode->getNodeType() === NodeInterface::TYPE_TRANSVERSE? GeneralNodesPanelStrategy::ROLE_ACCESS_GENERAL_NODE:TreeNodesPanelStrategy::ROLE_ACCESS_CREATE_NODE;
+        $this->denyAccessUnlessGranted($creationRole, $parentNode);
```
Comment (itkg-alavieille) [inline BackofficeBundle/Controller/NodeController.php]: why not use $node ?

File: BackofficeBundle/Resources/views/Tree/tree-macros.html.twig [hunk 4]
```diff
@@ -1,33 +1,36 @@
...
+                            {{ _self.menu_nodes(element.child, name) }}
+                        {% endif %}
+                        {{ _self.new_node(node.nodeId, "node-unsortable", node) }}
+                    </ul>
+                {% endif %}
+            </li>
+        {% endif %}
     {% else %}
-        {{ _self.new_node('-', "ui-state-unsortable") }}
+        {{ _self.new_node('-', "ui-state-unsortable", null) }}
```
Comment (itkg-alavieille) [inline BackofficeBundle/Resources/views/Tree/tree-macros.html.twig]: what happen in this case with null

File: ApiBundle/Controller/NodeController.php (hunk 2 above)
Comment (Thiblef) [inline ApiBundle/Controller/NodeController.php]: Cause they don't support specific roles like normal pages.

File: BackofficeBundle/Controller/NodeController.php (hunk 3 above)
Comment (Thiblef) [inline BackofficeBundle/Controller/NodeController.php]: Cause we have to check the parent node creation role. The 'role_access_create_node' allows users to create a node inside the actual node right?

Comment (Thiblef) [inline BackofficeBundle/Controller/NodeController.php]: Cause we have to check the parent node creation role. The 'role_access_create_node' allows users to create a node inside the actual node right?
We can't create a node and test after if we have the right to create it.

File: BackofficeBundle/Resources/views/Tree/tree-macros.html.twig (hunk 4 above)
Comment (Thiblef) [inline BackofficeBundle/Resources/views/Tree/tree-macros.html.twig]: It checks the global role.

File: ApiBundle/Controller/NodeController.php [hunk 5]
```diff
@@ -66,11 +68,18 @@ public function showOrCreateAction(Request $request, $nodeId, $errorNode)
         $language = $request->get('language', $currentSiteDefaultLanguage);
         $siteId = $currentSiteManager->getCurrentSiteId();
         $node = $this->findOneNode($nodeId, $language, $siteId, $request->get('version'));
-
+        if (!$errorNode && $node) {
```
Comment (itkg-ngilain) [inline ApiBundle/Controller/NodeController.php]: I think the ``showOrCreate`` method is growing dramatically which is not good for readability. Can use some protected functions call to make it cleraer ? Maybe one for the  ``if (!$errorNode && $node)`` and one for if (!$node)

Comment (Thiblef) [inline ApiBundle/Controller/NodeController.php]: I tried your solution and it's not making the code clearer in my sense cause we have to pass many parameters to each method.

**Comment by Thiblef** [comment]
@itkg-alavieille @itkg-ngilain @nicolasThal @itkg-nanne

File: ApiBundle/Controller/NodeController.php [hunk 6]
```diff
@@ -132,6 +138,30 @@ public function duplicateAction(Request $request, $nodeId, $version = null)
     }
 
     /**
+     * @param  NodeInterface $node
+     * @return string
+     */
+    protected function getAccessRole($node) {
```
Comment (nicolasThal) [inline ApiBundle/Controller/NodeController.php]: psr

File: ApiBundle/Controller/NodeController.php [hunk 7]
```diff
@@ -132,6 +138,30 @@ public function duplicateAction(Request $request, $nodeId, $version = null)
     }
 
     /**
+     * @param  NodeInterface $node
+     * @return string
```
Comment (nicolasThal) [inline ApiBundle/Controller/NodeController.php]: put a new line before the @return tag

File: ApiBundle/Controller/NodeController.php [hunk 8]
```diff
@@ -132,6 +138,30 @@ public function duplicateAction(Request $request, $nodeId, $version = null)
...
+        }
+
+        return TreeNodesPanelStrategy::ROLE_ACCESS_TREE_NODE;
+    }
+
+    /**
+     * @param  NodeInterface $node
+     * @return string
+     */
+    protected function getEditionRole($node) {
```
Comment (nicolasThal) [inline ApiBundle/Controller/NodeController.php]: same

File: ApiBundle/Controller/NodeController.php [hunk 9]
```diff
@@ -41,7 +41,9 @@ public function showAction(Request $request, $nodeId)
         $currentSiteDefaultLanguage = $currentSiteManager->getCurrentSiteDefaultLanguage();
         $language = $request->get('language', $currentSiteDefaultLanguage);
         $siteId = $currentSiteManager->getCurrentSiteId();
+        /** @var NodeInterface $node */
```
Comment (itkg-ngilain) [inline ApiBundle/Controller/NodeController.php]: You should use the interface on top of the file.

File: ApiBundle/Controller/NodeController.php [hunk 10]
```diff
@@ -66,11 +66,14 @@ public function showOrCreateAction(Request $request, $nodeId, $errorNode)
         $language = $request->get('language', $currentSiteDefaultLanguage);
         $siteId = $currentSiteManager->getCurrentSiteId();
         $node = $this->findOneNode($nodeId, $language, $siteId, $request->get('version'));
-
-        if (!$node) {
+        if (!$errorNode && $node) {
+            $this->denyAccessUnlessGranted($this->getAccessRole($node));
+        }
+        if(!$node){
```
Comment (itkg-ngilain) [inline ApiBundle/Controller/NodeController.php]: it lacks two spaces here

File: ApiBundle/Controller/NodeController.php [hunk 11]
```diff
@@ -276,4 +284,32 @@ protected function findOneNode($nodeId, $language, $siteId, $version = null)
 
         return $node;
     }
+
+    /**
+     * @param  NodeInterface $node
+     *
+     * @return string
+     */
+    protected function getAccessRole($node)
```
Comment (itkg-ngilain) [inline ApiBundle/Controller/NodeController.php]: Type your param with NodeInterface

File: ApiBundle/Controller/NodeController.php [hunk 12]
```diff
@@ -276,4 +284,32 @@ protected function findOneNode($nodeId, $language, $siteId, $version = null)
...
+
+        return TreeNodesPanelStrategy::ROLE_ACCESS_TREE_NODE;
+    }
+
+    /**
+     * @param  NodeInterface $node
+     *
+     * @return string
+     */
+    protected function getEditionRole($node)
```
Comment (itkg-ngilain) [inline ApiBundle/Controller/NodeController.php]: Same here

**Comment by itkg-ngilain** [comment]
:+1:

**Comment by SOLenG** [comment]
:+1:

**Comment by Thiblef** [comment]
@nicolasThal @itkg-alavieille @itkg-ngilain @SOLenG  PR updated
The solution chosen in Ceremony produce an access denied error when the user create a new version if : he has the general update role and not the update role on the node.

**Comment by Thiblef** [comment]
@itkg-ngilain @nicolasThal @itkg-nanne PR updated

File: ApiBundle/Controller/NodeController.php [hunk 13]
```diff
@@ -95,14 +99,15 @@ public function showOrCreateAction(Request $request, $nodeId, $errorNode)
...
-     * @Config\Security("is_granted('ROLE_ACCESS_DELETE_NODE')")
-     *
      * @return Response
      */
     public function deleteAction($nodeId)
     {
         $siteId = $this->get('open_orchestra_backoffice.context_manager')->getCurrentSiteId();
-        $nodes = $this->get('open_orchestra_model.repository.node')->findByNodeAndSite($nodeId, $siteId);
+        $nodes = $this->get('open_orchestra_model.repository.node')->findByNodeAndSiteSortedByVersion($nodeId, $siteId);
+        $node = !empty($nodes) ? $nodes[0] : null;
```
Comment (nicolasThal) [inline ApiBundle/Controller/NodeController.php]: array_pop could be working ?

Comment (Thiblef) [inline ApiBundle/Controller/NodeController.php]: array_shift could work. I'm going to change it

**Comment by nicolasThal** [comment]
I would agree with your pr if the functional test run on your local after a rebase

**Comment by Thiblef** [comment]
@nicolasThal Functional tests ok after a rebase.

**Comment by itkg-ngilain** [comment]
:+1:

**Comment by nicolasThal** [comment]
:+1:

**Pull request merged**